### DISCUSSION TITLE
Timed HLS cue points from JSON file

### DISCUSSION THREAD

**Discussion by Oliver-Mardle** [savonet/liquidsoap#4095] [General]
I'm currently looking into the insert_tag feature to add some custom tags to a HLS manifest.
I would like to read in a JSON file, and insert tags at the times stored in this file. What's the best method for doing this without using all of the system resources? I have figured out the JSON parsing, but now I need some sort of loop to compare the current time to the stored times in the JSON, and insert a tag if there's a match.

Previously, I had an external script triggered by on_file_change, however I now want to insert the tags using liquidsoap for better time accuracy, as it has the ability to end the current segment, insert a tag & start a new segment.

An example HLS manifest can be seen below with the CUE-OUT tag, currently inserted by an external script.

```
#EXT-X-CUE-OUT:120
#EXTINF:5.040,
aac_2_1723712344.83_117734.ts
#EXTINF:5.040,
aac_2_1723712349.65_117735.ts
#EXTINF:5.040,
aac_2_1723712354.62_117736.ts
#EXTINF:5.040,
aac_2_1723712359.72_117737.ts
#EXTINF:5.040,
aac_2_1723712364.94_117738.ts
#EXTINF:5.040,
aac_2_1723712369.88_117739.ts
#EXTINF:5.040,
aac_2_1723712374.95_117740.ts
```

**Comment by toots** [comment]
Hi!

This is a great feature. Our current approach with HLS is that we might not be able to include the full set of features but want to have entry points for users to add their own. I think that your need fits perfectly into that.

Currently, we provide extra tags options at the main and stream playlist level but they are static. I looks to me that you need a more dynamic way to insert tags, i.e. a `insert_tags` callback that could optionally also end the current segment.

Is that correct? If so, this is currently not supported. However, I should be able to add something to the `main`/`rolling-release-v2.3.x` pretty quickly.

**Reply by Oliver-Mardle** [reply]
An "insert_tags" callback would be absolutely perfect for our use. 
Yes please, if this could be added to the next release that would be fantastic! Thank you.

**Reply by toots** [reply]
Well turns out we already have it.. 😆

```liquidsoap
o = output.file.hls(
  [("mp3", %mp3), ...]
   ...
)

streams = o.streams()
let (_, mp3_stream) = list.find(fun s -> fst(s) == "mp3", streams)

# Insert tag:
mp3_stream.insert_tag("X-FOO")
```

(code untested but that's the idea).

Notes:
* Inserting tag is synchronous and will abort the current segment
* We don't support dynamic insertion of tags into the main playlist. I think that this is okay, I don't think the main playlist is supposed to be changed
* I'm adding a top-level `insert_tag` method that will iterate through all streams:

```liquidsoap
o.insert_tag("X-SAME-TAG-ON-ALL-STREAMS")
```

**Reply by Oliver-Mardle** [reply]
This is brilliant - I'll have a play around with this and see what I can get it to do. Thank you @toots
